### PR TITLE
Feature/mv 44 ability to choose camera and mic

### PR DESCRIPTION
**Pull request by kamil-stasiak**
## Goals

- At most one permission prompt when someone opens the page.
- If possible start the last selected device.
- Single prompt for audio and video.
- If the algorithm can’t start the last selected device (corrupted data, disconnected device) it should start any device.
- Should work on:
    - Desktop: Safari, Chrome, Firefox
    - IOS: Safari, Chrome
    - Android: Chrome

### Firefox

Firefox can show a permission manager window where the user can select one particular device. This window should be displayed only if no previous device info is present. Otherwise, we should ask for a specific device (prompt without a select input).

### Safari

Safari changes the device id for every session. Therefore, on Safari, I start the devices twice - once to fetch all the information, and once to select the device by its label.   

## Why everything is in one place?

My first attempt was to create specialized hooks for `navigator.mediaDevices?.enumerateDevices()` and `navigator.mediaDevices.getUserMedia()` but `enumerateDevices()`  does not properly work without `getUserMedia()`. To achieve optimal results, I have to combine invocations of these two methods. Believe me or not this code is much more readable than my previous attempt where enumerating devices and starting a new one has its own specialized hooks. It’s easier to read and understand an even long algorithm than an algorithm that involves the coordination of many hooks. 

After some time we should extract `assets/src/features/devices/useUserMedia.tsx` to `browser-media-utils` repository.

## Testing suggestions

The last selected device are stored in local storage. To test new user behavior clean up your local storage and reset permission in your browser. You can also corrupt data in local storage (deviceId and label are important) to test situations where the last device is not present. 

## Known limitations

If a new device is connected after the page loads it won’t be visible in the devices list. (`devicechange` event is not handled [https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/devicechange_event](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/devicechange_event))

### CODE REVIEW

File: assets/src/pages/room/components/MediaControlButton.tsx
Comment (pkrucz00) [inline assets/src/pages/room/components/MediaControlButton.tsx:45]: Just as above

File: assets/src/pages/room/components/MediaControlButton.tsx
Comment (pkrucz00) [inline assets/src/pages/room/components/MediaControlButton.tsx:24]: Maybe a default argument value would be enough?

**Review by pkrucz00 — CHANGES_REQUESTED** [review]

File: assets/src/contexts/LocalPeerContext.tsx
Comment (ostatni5) [inline assets/src/contexts/LocalPeerContext.tsx:32]: Consider changing names to `VIDEO_ID_KEY` and `AUDIO_ID_KEY`.

File: assets/src/contexts/LocalPeerContext.tsx
Comment (ostatni5) [reply assets/src/contexts/LocalPeerContext.tsx:51]: We should decide on `key` or `name`

**Review by bblaszkow06 — COMMENTED** [review]
Nothing to add on top of the existing comments, I'll wait for fixes before final approval :wink:

File: assets/src/features/home-page/components/VideoroomHomePage.tsx
Comment (kamil-stasiak) [reply assets/src/features/home-page/components/VideoroomHomePage.tsx:94]: Moved here: https://github.com/membraneframework/membrane_videoroom/pull/161/files#diff-d5115edf405b21061dc2186ea584b023dfbe28c983c0e503d6de0a23e729d1a3

**Review by pkrucz00 — CHANGES_REQUESTED** [review]
The PR is so big I'll try to divide the review like [a jellyfish hero once suggested](https://membraneframework.atlassian.net/wiki/spaces/MEMBRANEFR/pages/1343564/Some+thoughts+on+reviewing+code) that is:
1. Does it work and look properly. Since we do not have tests, a description about possible errors should be provided. 
2. Is the infrastructure (custom components, hooks, external libraries etc.) well organised and concise.
3. *Prettyfying* 🌈✨🦄 the codebase

I will probably group them into the ones that are about the popup and about the 
We'll se how it will turn out.

*The first part (chrome and mac) of the first part (api and design)*
Things that do not look or work as expected on Chrome on Mac:
  1. Sometimes my bluetooth headphones are completely silent if I open videoroom (i.e. when I listen to spotify, opening VR silence but doe not stop te music) and it did not used to be that way. Can you explain why it is so? Should it be a separate issue?   [It occurred only once so maybe it's neglectable].
  2. Screenshot 2023-04-26 at 13 32 29 Should the device id be visible? On chrome it is only on the video input device list.
  3.  For some reason the list is not responsive in terms of enabling or disabling the devices. When I plugged in a new camera, it had not appeared on the list. The same goes for disabling it (the unplugged audio device was still there even when I joined he room - I have no idea from where and what  the app have been sending then).
  4.   The popup item is not styled properly:
      a) title font is different and the font weight too small
      b) the list corners have been rounded too much
      c) The `Cancel` button should have red font
      d) I can't traverse the list with arrow keys when it is expanded
      e) cannot expand the list with enter (only with the spacebar)
      
The 3. point is in my opinion the most important. Please give it a look.

File: assets/src/App.tsx
Comment (kamil-stasiak) [reply assets/src/App.tsx:32]: When we reach all 9 of Dante's circles of hell it will be time to use some kind of state management library.

File: assets/package.json
Comment (kamil-stasiak) [reply assets/package.json]: It was released a few weeks ago 🚀

**Comment by kamil-stasiak** [comment]
> 1. Sometimes my bluetooth headphones are completely silent if I open videoroom (i.e. when I listen to spotify, opening VR silence but doe not stop te music) and it did not used to be that way. Can you explain why it is so? Should it be a separate issue? [It occurred only once so maybe it's neglectable].

🤔 I'll think about it. 

> 2. Should the device id be visible? On chrome it is only on the video input device list.

I've created jira ticket for that. We have to discuss this topic.

> 3. For some reason the list is not responsive in terms of enabling or disabling the devices. When I plugged in a new camera, it had not appeared on the list. The same goes for disabling it (the unplugged audio device was still there even when I joined he room - I have no idea from where and what the app have been sending then).

It is not implemented. I've create jira ticket for that.

> The popup item is not styled properly:
a) title font is different and the font weight too small
b) the list corners have been rounded too much
c) The Cancel button should have red font
d) I can't traverse the list with arrow keys when it is expanded
e) cannot expand the list with enter (only with the spacebar)

I'll fix that.

File: assets/src/features/devices/DeviceErrorBoundary.tsx
Comment (kamil-stasiak) [reply assets/src/features/devices/DeviceErrorBoundary.tsx:20]: Description added

**Comment by pkrucz00** [comment]
Continuing layout problems:
1. Screenshot 2023-04-27 at 10 04 01 The scroll on right does not look good. You can ~~steal~~ get inspired by the [membrane.steam](membrane.stream/contact) site where a similar problem occurs. 
Screenshot 2023-04-27 at 10 07 11
2. For some reason the audio is still audible when two microphones are muted. Are you sure muting the device did not change or still works properly?

3. A separate issue:  User on safari does not see other users are muted 
Screenshot 2023-04-27 at 10 12 10

File: assets/src/features/devices/useUserMedia.tsx
Comment (pkrucz00) [reply assets/src/features/devices/useUserMedia.tsx:299]: Nit: Could you ponder about extracting more functions like this?

File: assets/src/features/devices/useUserMedia.tsx
Comment (kamil-stasiak) [reply assets/src/features/devices/useUserMedia.tsx:595]: value of `refetchOnMount` is only used in "on mount" phase so every future change can be ignored

File: assets/src/features/devices/useUserMedia.tsx
Comment (kamil-stasiak) [reply assets/src/features/devices/useUserMedia.tsx]:
```ts
 const result = isDeviceDifferentFromLastSession(
      {
        deviceId: "abc",
        label: "xyz",
        kind: "audioinput",
        groupId: "",
        toJSON: () => {
          //
        },
      },
      {
        deviceId: "abc",
        label: "xyz",
        kind: "audioinput",
        groupId: "",
        toJSON: () => {
          //
        },
      }
    );

    console.log({ isDeviceDifferentFromLastSession: result }); // false
``

File: assets/src/features/devices/useUserMedia.tsx
Comment (kamil-stasiak) [reply assets/src/features/devices/useUserMedia.tsx:132]: It couldn't be made optional because it is not the last parameter. However, I don't want to make it optional because `undefined` is the appropriate value like `MediaTrackConstraints`

image

File: assets/src/features/devices/useUserMedia.tsx
Comment (kamil-stasiak) [reply assets/src/features/devices/useUserMedia.tsx:358]: From https://react.dev/reference/react/useRef
> useRef is a React Hook that lets you reference a value that’s not needed for rendering.

In this case I want to guard `init` method from additional invocations during rerender, so I used this mutable variable because changing this variable does not invoke rerender

File: assets/src/features/devices/useUserMedia.tsx
Comment (kamil-stasiak) [reply assets/src/features/devices/useUserMedia.tsx]: I'm not sure how to fix it since there are a lot of impurities in these lines. For now, I'll leave it as is.

File: assets/src/features/devices/useUserMedia.tsx
Comment (kamil-stasiak) [reply assets/src/features/devices/useUserMedia.tsx]: Fixed

File: assets/src/features/devices/useUserMedia.tsx
Comment (kamil-stasiak) [reply assets/src/features/devices/useUserMedia.tsx]: Some fixed

File: assets/src/features/devices/useUserMedia.tsx
Comment (pkrucz00) [reply assets/src/features/devices/useUserMedia.tsx]: This has been changed to:

```typescript
export type DeviceState = {
  status: DeviceReturnType;
  media: Media | null;
  error: DeviceError | null;
  devices: MediaDeviceInfo[] | null;
};

export type UseUserMediaState = {
  video: DeviceState;
  audio: DeviceState;
  devices: MediaDeviceInfo[] | null;
};
```

I do not understand, why the `devices` attribute is duplicated. How does it differ in both types?

Comment (kamil-stasiak) [reply assets/src/features/devices/useUserMedia.tsx]: Done

File: assets/src/features/devices/useUserMedia.tsx
Comment (pkrucz00) [reply assets/src/features/devices/useUserMedia.tsx]: Yeah, they can be pretty tricky due to many possible `undefine`s and small differences between video and audio data. It doesn't hurt much and can be ignored in most cases.

**Review by bblaszkow06 — APPROVED** [review]

File: assets/src/features/devices/use-user-media/useUserMedia.tsx
Comment (kamil-stasiak) [reply assets/src/features/devices/use-user-media/useUserMedia.tsx:279]: this one should be logged because it basically means that my assumptions are incorrect 💥

File: assets/src/features/shared/components/Select.tsx
Comment (pkrucz00) [inline assets/src/features/shared/components/Select.tsx:1]: Is this file taken from the www 🪼 website? If not, it should be there (@ostatni5 recently had to use something similar if I remember correctly, so this component would be used immediately).

**Review by pkrucz00 — COMMENTED** [review]
Beside some questions, I think this PR is good to go!
Those questions are in the unresolved comments. Please, take a look.

Gargantuan work has been done here. Good job!

File: assets/src/features/devices/disableSafariCache.ts
Comment (kamil-stasiak) [reply assets/src/features/devices/disableSafariCache.ts:2]: There it is: https://github.com/membraneframework/membrane_videoroom/pull/161/files#diff-6343522eec5fc04858831079e577b8d66a86ad16dda563f7230bef58a48b9578

**Comment by kamil-stasiak** [comment]
> In the context of debugging:
When I denied permission for both camera and microphone log has fired thrice. Should this be like this?
I think it should not hold this PR from being merged. Just a comment for future development.

Yes, this is exception based algorithm. In the worst case scenario, up to 7 exceptions can be thrown.

File: assets/src/features/devices/use-user-media/device-utils.ts
Comment (kamil-stasiak) [reply assets/src/features/devices/use-user-media/device-utils.ts:65]: If make that change have to invoke this function with additional context `enumerateDevices.apply(navigator.mediaDevices);`

File: assets/src/features/devices/DeviceErrorBoundary.tsx
Comment (kamil-stasiak) [reply assets/src/features/devices/DeviceErrorBoundary.tsx:34]: I'll add jira ticket for that

File: assets/src/features/devices/DeviceErrorBoundary.tsx
Comment (kamil-stasiak) [reply assets/src/features/devices/DeviceErrorBoundary.tsx:12]: I'll add jira ticket for that

File: assets/src/features/devices/DeviceErrorBoundary.tsx
Comment (kamil-stasiak) [reply assets/src/features/devices/DeviceErrorBoundary.tsx:34]: I'll add jira ticket for that

File: assets/src/features/devices/use-user-media/useUserMedia.tsx
Comment (kamil-stasiak) [reply assets/src/features/devices/use-user-media/useUserMedia.tsx:287]: Yes, this is exception based algorithm. In the worst case scenario, up to 7 exceptions can be thrown.

File: assets/src/features/shared/components/Select.tsx
Comment (kamil-stasiak) [reply assets/src/features/shared/components/Select.tsx:1]: We will discuss sharing components in a separate task.

File: assets/src/contexts/LocalPeerContext.tsx
Comment (kamil-stasiak) [reply assets/src/contexts/LocalPeerContext.tsx]: Fixed

File: assets/src/pages/room/RoomPage.tsx
Comment (kamil-stasiak) [reply assets/src/pages/room/RoomPage.tsx]: Fixed

File: assets/src/features/home-page/components/HomePageVideoTile.tsx
Comment (kamil-stasiak) [reply assets/src/features/home-page/components/HomePageVideoTile.tsx]: fixed

File: assets/src/pages/room/hooks/useSetLocalUserTrack.tsx
Comment (kamil-stasiak) [reply assets/src/pages/room/hooks/useSetLocalUserTrack.tsx]: fixed

**Comment by kamil-stasiak** [comment]
> The scroll on right does not look good. You can steal get inspired by the [membrane.steam](https://github.com/membraneframework/membrane_videoroom/pull/membrane.stream/contact) site where a similar problem occurs.

We will fix scrollbar later :)

**Pull request merged**
